src/trajectory.jl

```diff
@@ -227,22 +227,52 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
 
 ### Multinomial sampling from trajecory
 
-function randcat(rng::AbstractRNG, xs, p)
-    u = rand(rng)
+function randcat(rng::AbstractRNG, p_unorm::AbstractVector{T}) where {T}
+    p = p_unorm ./ sum(p_unorm)
+    u = rand(rng, T)
     cp = zero(eltype(p))
     i = 0
     while cp < u
-        cp += p[i +=1]
+        cp += p[i+=1]
     end
-    return xs[max(i, 1)]
+    return max(i, 1)
+end
+
+function randcat(rng, P_unorm::AbstractMatrix{T}) where {T}
+    P = P_unorm ./ sum(P_unorm; dims=2)
+    u = rand(rng, T, size(P, 1))
+    C = cumsum(P; dims=2)
+    is = convert.(Int, vec(sum(C .< u; dims=2)))
+    return max.(is, 1)
+end
+
+randcat(rng::AbstractRNG, xs, p::AbstractVector) = xs[randcat(rng, p)]
+
+# xs is in the form of Vector{Vector} and has shape [n_steps][]
+function randcat(rng, xs, P_unorm::AbstractMatrix)
```

Comment:
Similiar here.
```suggestion
function randcat(rng, zs::VectorOrMat{PhasePoint}, P_unorm::AbstractMatrix)
```

Reply:
`zs` is always a vector of phase points. In vectorized mode, a single `z` stores all phase points for all chains for a particular step. I think this is a bit conterintuitive and I was trying to figure out a way to improve this. In particular, maybe we should all multiple steps to be sotred in a single `PhasePoint`. This would make a low of mutiple-chain/multiple-step operations easier. But I think it worth another PR with more discussions.

Reply:
> In particular, maybe we should all multiple steps to be sotred in a single PhasePoint. 

I'm not sure, saving more than one (position, momentum) pairs in a PhasePoint would make PhasePoint and leapfrog related code complicated. Maybe there is a better approach.

Reply:
It will certainly make the code more complicated. But I think any solution have to do so, no matter in which manner, to achive high perfomrance.

Reply:
Perhaps we can keep this discussion for future PRs.